### PR TITLE
Switch to failsafe-go circuit breaker implementation

### PR DESCRIPTION
**Pull request by 56quarters**
#### What this PR does

Switch to failsafe circuit breaker implementation that allows us to define an error rate over a moving window instead new windows every N seconds. Helps for high traffic clusters where the long tail of requests might exceed the timeout enough in raw numbers but still very infrequently compared to request volume.

#### Which issue(s) this PR fixes or relates to

Related #5650 

#### Checklist

- [X] Tests updated
- [X] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by jhalterman — APPROVED** [review]

File: cmd/mimir/config-descriptor.json [hunk 1]
```diff
@@ -2116,20 +2116,42 @@
...
+              "desc": "How many requests must have been executed in period for the circuit breaker to be eligible to open for the rate of failures",
+              "fieldValue": null,
+              "fieldDefaultValue": 100,
+              "fieldFlag": "ingester.client.circuit-breaker.failure-execution-threshold",
+              "fieldType": "int",
+              "fieldCategory": "experimental"
+            },
+            {
+              "kind": "field",
+              "name": "period",
```
Comment (jhalterman) [inline cmd/mimir/config-descriptor.json:2139]: Maybe we could call this thresholding_period, or failure_thresholding_period?

Comment (56quarters) [reply cmd/mimir/config-descriptor.json:2139]: Sure, either of those works for me. Do you have a preference?

File: cmd/mimir/config-descriptor.json [hunk 2]
```diff
@@ -2116,20 +2116,42 @@
...
+              "required": false,
+              "desc": "Moving window of time that the percentage of failed requests is computed over",
+              "fieldValue": null,
+              "fieldDefaultValue": 10000000000,
```
Comment (56quarters) [inline cmd/mimir/config-descriptor.json:2143]: Based on offline discussion, I think you are right @jhalterman that a longer period here makes such that the default value works for both read and write paths. Looking at a handful of production cells, the lowest traffic one was doing ~15 `/cortex.Ingester/QueryStream` requests / sec / ingester. Changing this period to `1m` would ensure that all cells meet the "100 requests" threshold for circuit breaking on the read path.

File: cmd/mimir/config-descriptor.json (hunk 1 above)
Comment (56quarters) [reply cmd/mimir/config-descriptor.json]: Went with `thresholding_period`

**Comment by 56quarters** [comment]
To summarize the default behavior when circuit breaking is enabled for reviewers:

* If at least 100 requests have been executed over the last 1m and 10% of them have failed, open the circuit breaker
  * 100 requests comes from `failure-execution-threshold`
  * 10% comes from `failure-threshold`
  * 1m comes from `thresholding-period`
* After 1m has elapsed, allow 100 requests through to see if the ingester has recovered
  * 1m comes from `cooldown-period`
  * 100 requests comes from `failure-execution-threshold`
* If the failure rate for those 100 requests is under 10%, close the circuit breaker.

If anyone has concerns or thoughts about the defaults here, please LMK

File: cmd/mimir/config-descriptor.json [hunk 3]
```diff
@@ -2116,20 +2116,42 @@
               "kind": "field",
               "name": "failure_threshold",
               "required": false,
-              "desc": "Max number of requests that can fail in a row before the circuit breaker opens",
+              "desc": "Max percentage of requests that can fail over period before the circuit breaker opens",
               "fieldValue": null,
               "fieldDefaultValue": 10,
               "fieldFlag": "ingester.client.circuit-breaker.failure-threshold",
               "fieldType": "int",
```
Comment (charleskorn) [inline cmd/mimir/config-descriptor.json:2123]: If this field is a percentage now, should it be a float? (Would we want to be able to set the percentage at a more granular level than 1%?)

**Review by charleskorn — APPROVED** [review]
LGTM, one non-blocking comment

File: cmd/mimir/config-descriptor.json (hunk 3 above)
Comment (jhalterman) [reply cmd/mimir/config-descriptor.json:2123]: It's an int, from 0-100 (the precedent for doing this with circuitbreakers came from [Hystrix](https://github.com/Netflix/Hystrix/wiki/Configuration#circuitBreaker.errorThresholdPercentage), FWIW). I'm sure 1% increments are granular enough. A harder problem is choosing a good value, but even a default like 10% is probably better than guessing at rate limits or concurrency limits.

Comment (56quarters) [reply cmd/mimir/config-descriptor.json:2123]: We could make it a float but the underlying library (failsafe) uses a `uint` from 0 to 100 so we'd have the same granularity at the end of the day.

**Pull request merged**